theano/tensor/nnet/corr_gemm.c

```diff
@@ -0,0 +1,357 @@
+              N_, M_, K_,
+              1.,
+              col_data + col_offset * g, N_,
+              weight + weight_offset * g, K_,
+              0.,
+              top_data + (*top)[i]->offset(n) + top_offset * g, N_);
+          }
+        }
+        */
+    }
+    else if (direction == 1) {  // backprop wrt. weights
+        output = weight;
+        // valid convolution: im2col, then gemm
+        // Iterate over batch
+        for (int n = 0; n < batchSize; n++) {
+            // First, im2col
+            im2col((%(float_type)s*)PyArray_DATA(bottom) + n * bottom_stride, nChannels, bottomHeight,
+                    bottomWidth, kH, kW, padH, padW, dH, dW, (%(float_type)s*)PyArray_DATA(col));
+            // Second, gemm
+            // Note that we accumulate into weight. We do so by setting beta = 0
+            // for the first iteration and beta = 1 for subsequent ones. (This
+            // is faster than setting weight to all zeros before the loop.)
+            %(gemm)s(&Trans, &NTrans,
+                   &K_, &M_, &N_,
+                   &one,
+                   (%(float_type)s*)PyArray_DATA(col), &N_,
+                   (%(float_type)s*)PyArray_DATA(top) + n * top_stride, &N_,
+                   (n == 0) ? &zero : &one,
+                   (%(float_type)s*)PyArray_DATA(weight), &K_);
+        }
+        /*
+        // Original caffe code for comparison
+        // Note that this code was translated from the Theano GPU code,
+        // not the Caffe CPU code.
+        // https://github.com/BVLC/caffe/blob/master/src/caffe/layers/conv_layer.cu
+        // Note that this is for grouped convolution; we can ignore groups
+        for (int n = 0; n < num_; ++n) {
+          // Since we saved memory in the forward pass by not storing all col
+          // data, we will need to recompute them.
+          im2col_gpu(bottom_data + (*bottom)[i]->offset(n), channels_, height_,
+                     width_, kernel_h_, kernel_w_, pad_h_, pad_w_,
+                     stride_h_, stride_w_, col_data);
+          // gradient w.r.t. weight. Note that we will accumulate diffs.
+          for (int g = 0; g < group_; ++g) {
+            caffe_gpu_gemm<Dtype>(CblasNoTrans, CblasTrans, M_, K_, N_,
+                (Dtype)1., top_diff + top[i]->offset(n) + top_offset * g,
+                col_data + col_offset * g, (Dtype)1.,
+                weight_diff + weight_offset * g);
+            == (see https://github.com/BVLC/caffe/blob/master/src/caffe/util/math_functions.cu#L16)
+            cublasSgemm(CUBLAS_OP_T, CUBLAS_OP_N, K_, M_, N_,
+                1.0,
+                col_data + col_offset * g, N_,
+                top_diff + top[i]->offset(n) + top_offset * g, N_,
+                1.0,
+                weight_diff + weight_offset * g, K_);
+          }
+        }
+        */
+    }
+    else if (direction == 2) {  // backprop wrt. inputs
+        output = bottom;
+        PyArray_FILLWBYTE(bottom, 0);
```

Comment:
It's mostly just a result of the caffe col2im and the available calls in the numpy C-API.

The ```col2im_cpu``` function that caffe uses calls ```caffe_set``` on one slice of the output array:
https://github.com/BVLC/caffe/blob/4f8775c1e88604879edef0698acab7b0d8a2ab43/src/caffe/util/im2col.cpp#L53
I looked through the numpy C-API and there didn't seem to be an equivalent call.

And in the GPU version you worked on, the result at each output location is accumulated in each thread (```val```) and then the value in the output array is overwritten. To do this on CPU would require allocating an extra array and accumulating into it, which seems more complicated.

Since col2im won't be used outside of this file, setting the array to zero outside of col2im seemed reasonable. I'll add a comment though so it's clear why that is happening.